# [Model Evaluation Metric] Classifier Model 

## 1. Confusion Matrix
### 1) Accuracy (정확도)
Accuracy in classification problems is the number of correct predictions made by the model over all kinds predictions made

분류 문제(Classification problem)에서 Accuracy는 모든 예측값에 대하여 올바르게 예측된 값을 나타냅니다.

Accuracy는 양성을 양성이라고 예측하고 음성을 음성이라고 예측한 확률을 말합니다.

즉, 예측 결과가 실제 값과 얼마나 정확하게 맞는 지를 평가하는 지표


#### 1. sklearn에서 제공하는 classifier 객체의 score 메서드
- classifier_model.score(X_test, y_test) # Return the mean accuracy on the given test data and labels

#### 2. sklearn.metrics의 accuracy_score 메서드를 사용하면 accuracy score를 구할 수 있습니다
- accuracy_score(y_test, preds)
- from sklearn.metrics import accuracy_score 를 해주어야 합니다.
- Accuracy is good measure when the target variable class in the data are nearly balanced. example Survived(60%-yes, 40% no)
- Accuracy score는 target variable이 거의 균등할때 사용하면 좋습니다. 

In [18]:
import platform
import matplotlib
import seaborn as sns 
import matplotlib.pyplot as plt 
%matplotlib inline


In [19]:
# 01 데이터 셋 불러오기
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

cancer = load_breast_cancer()
X = cancer.data  #입력
y = cancer.target # 출력 

In [21]:
# f, ax = plt.subplots(1, 2, figsize=(12, 4)) 
# y.value_counts().plot.pie(autopct='%1.1f%%', ax=ax[0]) 
# sns.countplot(y, data=X_, ax=ax[1])

In [4]:
# 02 데이터 셋 나누기 및 학습
X_train, X_test, y_train, y_test = train_test_split(X,y ,
                                                    stratify=cancer.target, random_state=42)

1. sklearn에서 제공하는 classifier 객체의 score 메서드

In [6]:
# 03 트리5개로 구성된 랜덤 포레스트 모델 만들기
model = RandomForestClassifier(n_estimators=5, random_state=0) # 5개의 트리
model.fit(X_train, y_train)  #모델 학습

print("학습용 세트 정확도 : {:.3f}".format(model.score(X_train, y_train))) # score가 분류문제에서는 정확도가 찍히고, 회기문제에서는 결정도가 찍힘
print("테스트 세트 정확도 : {:.3f}".format(model.score(X_test, y_test)))  #



#0.973 의 결과는 테스트세트에서 97.3% 를 정확히 예측했다는 의미이다. (실행할 때마다 결과가 달라진다)
#train data가 test data보다 성능이 더 좋음 > 오버피팅이 되어 있음.
#RandomForestClassifier의 parameter를 조정 ()

학습용 세트 정확도 : 1.000
테스트 세트 정확도 : 0.958


In [8]:
# predict()는 테스트 데이터의 분류 결과를 예측합니다. (회귀에서는>> 반환값이 예측 확률입니다.)
pred = model.predict(X_test)
pred

array([1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1,
       1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1,
       1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1])

2. sklearn.metrics의 accuracy_score 메서드를 사용하면 accuracy score

In [10]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, pred)

0.958041958041958

### 2) f1-score
F1 score is the weighted average of Precision and Recall.   
Therefore, this score takes both false positives and false negatives into account.  
Intuitively it is not as easy to understand as accuracy,   
but F1 is usually more useful than accuracy, especially if you have an uneven class distribution

- f1-score는 Precision과 Recall에 대한  가중 평균(weighted average)입니다.
- f1-score는 데이터가 imbalance할 때 측정 도구로 사용하면 좋습니다
- 우리가 마주치는 대부분의 데이터는 불균형한 데이터의 분포를 갖기에 f1-scroe를 알아두시면 모델을 평가하는데 많은 도움이 됩니다.

# Metrics 출력

In [14]:
from sklearn.metrics import classification_report 
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.98      0.91      0.94        53
           1       0.95      0.99      0.97        90

    accuracy                           0.96       143
   macro avg       0.96      0.95      0.95       143
weighted avg       0.96      0.96      0.96       143



모델의 정확도는 96%입니다. 

f1-score를 보면 macro, weigted가 존재합니다. 두 값의 차이는 무엇일까요?

macro avg: 단순평균값
weighted avg: 각 클래스에 속하는 표본의 갯수로 가중평균값
다시 한 번 해석해보면 우리의 모델은  단순평균값으로는 95%의 f1-score를 보이며 가중평균으로는 0.96의 f1-score를 나타냅니다.

Accuracy와도 크게 차이가 나지않는것 보면 클래스 불균형으로 인한 문제는 적어보입니다.



출처: https://deepinsight.tistory.com/173 [Steve-Lee's Deep Insight]